types/src/v0/header.rs

```diff
@@ -136,50 +170,54 @@ impl<'de> Deserialize<'de> for Header {
             where
                 V: MapAccess<'de>,
             {
-                // retrieving the first entry from the map, which could either be a
-                // `chain_config` (for v0.1 headers) or a `version` (for `VersionedHeader` structs).
-                let Some((_, chain_config_or_version)) =
-                    map.next_entry::<StructFields, ResolvableChainConfigOrVersion>()?
-                else {
-                    return Err(de::Error::missing_field("version"));
-                };
+                // remaining v1 header fields are inserted in the hashmap
+                let mut v1_hashmap: HashMap<StructFields, Value> = HashMap::new();
+                let mut chain_config_or_version = None;
+                let mut fields: Option<Value> = None;
+
+                while let Some(key) = map.next_key::<StructFields>()? {
+                    match key {
+                        StructFields::ChainConfig => {
+                            chain_config_or_version = Some(map.next_value()?)
+                        }
+                        StructFields::Version => chain_config_or_version = Some(map.next_value()?),
+                        StructFields::Fields => fields = Some(map.next_value()?),
+                        _ => {
+                            v1_hashmap.insert(key, map.next_value()?);
+                        }
+                    }
+                }
+
+                let chain_config_or_version: ResolvableChainConfigOrVersion =
+                    chain_config_or_version
+                        .ok_or_else(|| de::Error::missing_field("chain_config_or_version"))?;
 
                 // Match to determine the appropriate header version.
                 match chain_config_or_version.chain_config {
                     EitherOrVersion::Left(cfg) => Ok(Header::V1(
-                        v0_1::Header::deserialize_with_chain_config_map(cfg.into(), map)?,
+                        v0_1::Header::deserialize_with_chain_config_map::<V>(
+                            cfg.into(),
+                            v1_hashmap,
+                        )?,
                     )),
                     EitherOrVersion::Right(commit) => Ok(Header::V1(
-                        v0_1::Header::deserialize_with_chain_config_map(commit.into(), map)?,
+                        v0_1::Header::deserialize_with_chain_config_map::<V>(
+                            commit.into(),
+                            v1_hashmap,
+                        )?,
                     )),
                     EitherOrVersion::Version(Version { major: 0, minor: 2 }) => {
-                        Ok(Header::V2(map.next_entry::<StructFields, _>()?.unwrap().1))
+                        Ok(Header::V2(serde_json::from_value(fields.unwrap()).unwrap()))
```

Comment:
Don't unwrap. If `fields` is missing or if it doesn't deserialize into the expected type, it is a reportable error